src/models/room.rs

```diff
@@ -0,0 +1,373 @@
+            alias.alias().to_string()
+        } else if !self.aliases.is_empty() {
+            self.aliases[0].alias().to_string()
+        } else {
+            // TODO
+            let mut names = members
+                .values()
+                .flat_map(|m| m.user.display_name.clone())
+                .take(3)
+                .collect::<Vec<_>>();
+
+            if names.is_empty() {
+                // TODO implement the rest of matrix-js-sdk handling of room names
+                format!("Room {}", room_id)
+            } else {
+                // stabilize order
+                names.sort();
+                names.join(", ")
+            }
+        }
+    }
+}
+
+impl Room {
+    /// Create a new room.
+    ///
+    /// # Arguments
+    ///
+    /// * `room_id` - The unique id of the room.
+    ///
+    /// * `own_user_id` - The mxid of our own user.
+    pub fn new(room_id: &str, own_user_id: &str) -> Self {
+        Room {
+            room_id: room_id.to_string(),
+            room_name: RoomName::default(),
+            own_user_id: own_user_id.to_owned(),
+            creator: None,
+            members: HashMap::new(),
+            typing_users: Vec::new(),
+            encrypted: false,
+        }
+    }
+
+    fn add_member(&mut self, event: &MemberEvent) -> bool {
+        if self.members.contains_key(&event.state_key) {
+            return false;
+        }
+
+        let member = RoomMember::new(event);
+
+        self.members.insert(event.state_key.clone(), member);
+
+        true
+    }
+
+    // fn remove_member(&mut self, event: &MemberEvent) -> bool {
+    //     if let Some(member) = self.members.get_mut(&event.sender.to_string()) {
+    //         let changed = member.membership == event.content.membership;
+    //         member.membership = event.content.membership;
+    //         changed
+    //     } else {
+    //         false
+    //     }
+    // }
+
+    // fn update_joined_member(&mut self, event: &MemberEvent) -> bool {
+    //     if let Some(member) = self.members.get_mut(&event.state_key) {
+    //         member.update(event);
+    //     }
+
+    //     false
+    // }
+
+    // fn handle_join(&mut self, event: &MemberEvent) -> bool {
+    //     match &event.prev_content {
+    //         Some(c) => match c.membership {
+    //             MembershipState::Join => self.update_joined_member(event),
+    //             MembershipState::Invite => self.add_member(event),
+    //             MembershipState::Leave => self.remove_member(event),
+    //             _ => false,
+    //         },
+    //         None => self.add_member(event),
+    //     }
+    // }
+
+    // fn handle_leave(&mut self, event: &MemberEvent) -> bool {
+        
+    // }
+
+    /// Handle a room.member updating the room state if necessary.
+    /// 
+    /// Returns true if the joined member list changed, false otherwise.
+    pub fn handle_membership(&mut self, event: &MemberEvent) -> bool {
```

Comment:
Is this ok, I removed the `handle_*` methods except `add_member` since it now only updates the status, I'll have to make sure any code that maps over the members knows to check membership status.

Reply:
Ruma-events recently got [this](https://docs.rs/ruma-events/0.17.0/ruma_events/room/member/struct.MemberEvent.html#method.membership_change) method. So we don't need to figure out the diff ourselves.

Returning a bool if the memberlist changes will be important for encryption since we need to send keys out to new members or expire them if members left.

What do you mean by only updates the status? Sorry didn't yet got in detail through the code.

Reply:
I think what I meant was that now that the `MembershipState` information is kept the implementation for each `handle_*` method is the same with the exception of `add_member`?

> Returning a bool if the memberlist changes will be important for encryption since we need to send keys out to new members or expire them if members left.

Ahh that makes sense, I will be sure to have any state mutating method follow `foo() -> bool`

Reply:
Ah, that's because most of the handle_* methods are stubs, they would need to do something interesting. If a user joins we add them to the room state, if a user leaves we remove them.

Reply:
It's fine to remove them now, but the functionality should eventually be added.

Reply:
Ok so like the python version I should just remove the `RoomMember` instead of updating the `MembershipState` and leaving the user in the HashMap (which is what I am doing now) [here](https://github.com/DevinR528/matrix-rust-sdk/blob/room-name/src/models/room_member.rs#L81-L85)?

Reply:
That's a good question, keeping them there would mean increased memory usage.

Though am I forgetting any downsides of not keeping them around?

Reply:
Keeping them around may simplify dealing with banned and invited but not joined users. Simplify in the sense of using one struct to represent all `RoomMember` states, but as you said it may bloat and complicate the iterating through and use of the members `HashMap`.